accepted/exception-handling.md

```diff
@@ -0,0 +1,514 @@
+    jump out of signal handlers on traps, which is quite subtle and
+    tricky. That said, we could revisit this decision and investigate
+    sharing these mechanisms at some point in the future. It isn't
+    something we must grapple with now, so for simplicity and
+    expediency, we won't.
+
+For the concrete design of the three-way result type, there are
+multiple options to consider:
+
+1. A genuine three-parameter result type capturing the three ways in
+   which a Wasm program can terminate, e.g.
+
+   ```rust
+   pub enum ThreeWayResult<T> {
+       Ok(T),
+       Exception(wasmtime::Exception),
+       Trap(wasmtime::Trap),
+   }
+   ```
+
+   We would not be able to use `?` (because Rust's `Try` trait is
+   currently an unstable feature) and would need to define our own
+   `try!`-style macros. The resulting ergonomics for regular code that
+   just propagates errors and exceptions would be poor.
+
+2. A nested result type, e.g. `Result<Result<Success, Exception>,
+   Trap>` to remain compatible with Rust's `?` operator. The first
+   usage of `?` would propagate traps and, in theory, a second `?`
+   would propagate exceptions. In practice, our ability to actually
+   use a second `?` would be limited due to the combination of
+   mismatched levels of `Result` nesting and, again, Rust's `Try`
+   trait being an unstable feature. While this could sometimes provide
+   better ergonomics than option (1), it wouldn't be a large
+   improvement.
+
+3. We could define an error `enum` that is either a trap or an
+   exception:
+
+   ```rust
+   pub enum TrapOrException {
+       Trap(wasmtime::Trap),
+       Exception(wasmtime::Exception),
+   }
+   ```
+
+   We would then use `Result<Success, TrapOrException>` as the result
+   of running a Wasm program or calling into Wasm.
+
+   This would allow for good `?`-propagation ergonomics. It also
+   forces callers to differentiate between exceptions and traps, which
+   could be either a good or bad thing depending on context and how
+   often that is something callers need to care about. I think we
+   could largely alleviate this with a `From<TrapOrException>`
+   conversion implementation for `anyhow::Error`, or by implementing
+   `std::error::Error` for `TrapOrException` and piggybacking on
+   `anyhow`'s generic error-to-`anyhow::Error` conversion
+   implementation.
+
+4. The final option is to continue using `Result<Success,
+   anyhow::Error>` and allow callers to try to downcast the
+   `anyhow::Error` to either a `wasmtime::Trap` or a
+   `wasmtime::Exception` as needed.
+
+   This also allows for good `?`-propagation ergonomics. The primary
+   difference between this and option (3) is that it is a lot easier
+   to forget to check for exception versus trap, if that is something
+   a caller needs to do. If it isn't something the caller needs to do,
+   then it erases what would otherwise have been unnecessary cognitive
+   overhead.
```

Comment:
For the option (3) vs (4) category I'd vote for (4) and would say that we can alleviate this problem by we pretty much have a [single location that converts `Result<T>` into `T`](https://github.com/bytecodealliance/wasmtime/blob/132a490d72e6381c593a409aeca50a765b54ca28/crates/wasmtime/src/runtime/vm/traphandlers.rs#L215) since as the host we also have to do things like `catch_unwind` which we're doing there as well. Given that I'd vote for (4) as it's less churn in general.

Reply:
> we pretty much have a [single location that converts `Result<T>` into `T`](https://github.com/bytecodealliance/wasmtime/blob/132a490d72e6381c593a409aeca50a765b54ca28/crates/wasmtime/src/runtime/vm/traphandlers.rs#L215)

I meant guest code, not VM code. All this discussion is in the context of, e.g., what kind of result should we return from `Func::call`. While our VM code is fixed and we can make claims like "there is one place that cares about X", guest code is an open world where we don't know how often people want to check for, handle, or automatically propagate exception results from `Func::call`. If we knew that answer, then choosing between (3) and (4) would be straightforward. But we don't yet, and it is also somewhat future-looking, so I am not sure how to proceed.